src/lib/api-keys.ts

```diff
@@ -1,6 +1,74 @@
+import type { Request } from 'express';
 import * as _ from 'lodash';
+import { errors } from '@balena/pinejs';
+import {
+	createApiKey,
+	ApiKeyOptions,
+	PartialCreateKey,
+} from '../platform/api-keys';
+
+const { BadRequestError } = errors;
+
+const supportedActorTypes = ['application', 'user', 'device'] as const;
+
+export type ApiKeyParameters = {
+	actorType: typeof supportedActorTypes[number];
+	actorTypeId: number;
+	roles: string[];
+} & Pick<ApiKeyOptions, 'name' | 'description' | 'apiKey'>;
+
+export const createGenericApiKey = async (
+	req: Request,
+	{
+		actorType,
+		actorTypeId,
+		roles,
+		name,
+		description,
+		apiKey,
+	}: ApiKeyParameters,
+) => {
+	if (!supportedActorTypes.includes(actorType)) {
+		throw new BadRequestError('Unsupported actor type');
+	}
+
+	if (!Number.isFinite(actorTypeId)) {
+		throw new BadRequestError('Actor must be a number');
+	}
+
+	if (!Array.isArray(roles) || roles.some((r) => typeof r !== 'string' || !r)) {
```

Comment:
If you change this to `every` then it will also fail for an empty array, eg
```suggestion
	if (!Array.isArray(roles) || roles.every((r) => typeof r === 'string' && r.length > 0)) {
```
see:
```
[].every(() => false) // true

Reply:
I guess you missed the `!` in front of the `every`.

Thought of that but it felt that `some` is cleaner than `!every` (the `!` makes then equivalent, despite their 0 length difference in the general case).
PS: In my head `some` sounds like the early return pattern, and that's probably why I prefer it.

Updated a bit, but let me know if you still prefer `!every`.